chromium_src/chrome/browser/ssl/https_only_mode_navigation_throttle.cc

```diff
@@ -0,0 +1,143 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "chrome/browser/ssl/https_only_mode_navigation_throttle.h"
+
+#include "base/threading/sequenced_task_runner_handle.h"
+#include "base/time/time.h"
+#include "brave/browser/brave_browser_process.h"
+#include "brave/components/brave_shields/browser/brave_shields_util.h"
+#include "chrome/browser/content_settings/host_content_settings_map_factory.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ssl/https_only_mode_controller_client.h"
+#include "components/prefs/pref_service.h"
+#include "components/security_interstitials/content/stateful_ssl_host_state_delegate.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/page_navigator.h"
+#include "content/public/browser/web_contents.h"
+
+namespace content {
+class BrowserContext;
+}  // namespace content
+
+namespace {
+
+// Tor is slow and needs a longer fallback delay
+constexpr base::TimeDelta kTorFallbackDelay = base::Seconds(20);
+
+bool ShouldUpgradeToHttps(content::NavigationHandle* handle) {
+  if (!brave_shields::IsHttpsByDefaultFeatureEnabled()) {
+    return false;
+  }
+  content::BrowserContext* context =
+      handle->GetWebContents()->GetBrowserContext();
+  const GURL& url = handle->GetURL();
+  HostContentSettingsMap* map =
+      HostContentSettingsMapFactory::GetForProfile(context);
+  return brave_shields::ShouldUpgradeToHttps(
+      map, url, g_brave_browser_process->https_upgrade_exceptions_service());
+}
+
+bool IsTor(content::NavigationHandle* handle) {
+  auto* context = handle->GetWebContents()->GetBrowserContext();
+  Profile* profile = Profile::FromBrowserContext(context);
+  return profile->IsTor();
+}
+
+}  // namespace
+
+#define WillFailRequest WillFailRequest_ChromiumImpl
+#define GetBoolean(PREF_NAME) \
+  GetBooleanOr(PREF_NAME, ShouldUpgradeToHttps(handle))
+#define SetNavigationTimeout(DEFAULT_TIMEOUT)                         \
+  SetNavigationTimeout(IsTor(navigation_handle()) ? kTorFallbackDelay \
```

Comment:
`IsTor` is called and this change is applied regardless of the enabled feature state. is this really what we want?

Reply:
I believe this is OK, because this code only runs when we know that the navigation has been upgraded.